snooty/diagnostics.py

```diff
@@ -0,0 +1,204 @@
+    severity = Diagnostic.Level.error
+
+
+# this could captuer 'literal include expected path arg..as well?
+class ExpectedPathArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        name: Path,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f'"{name}" expected a path argument', start, end)
+        self.name = name
+
+
+class ExpectedImgArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class ImgExpectedButNotRequired(Diagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class OptionsNotSupported(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class GitMergeConflictArtifactFound(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        path: Path,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(
+            f"Git merge conflict artifact found in {str(path)} on line {str(start)}",
+            start,
+            end,
+        )
+        self.path = path
+
+
+class DocUtilsParseError(Diagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class ErrorParsingYAMLFile(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        path: Path,
+        reason: str,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f"Error parsing YAML file {str(path)}: {reason}", start, end)
+        self.path = path
+        self.reason = reason
+
+
+class InvalidLiteralInclude(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class SubstitutionRefError(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class VariableNotDeclaredConstant(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class InvalidTableStructure(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class MissingOption(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class MissingRef(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        name: str,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f"Missing ref; all {name} must define a ref", start, end)
+        self.name = name
+
+
+class FailedToInheritRef(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class RefAlreadyExists(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class UnknownSubstitution(Diagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class TargetNotFound(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class AmbiguousTarget(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class TodoInfo(Diagnostic):
+    severity = Diagnostic.Level.info
+
+
+class ErrorLoadingFile(Diagnostic):
```

Comment:
Sorry for not catching this in the spec, but what is the semantic difference between `ErrorLoadingFile` and `CannotOpenFile`? I can see that `ErrorLoadingFile` is specifically thrown when a `LoadError` is caught and `CannotOpenFile` for `OSError` and others, but perhaps there is no meaningful difference to the user and they're worth combining.

Reply:
I believe it was your call @i80and to have ErrorLoadingFile and then CannotOpenFile and for the to latter capture `CannotOpenFile`,  `FileDoesNotExist`, `CannotOpenLiteralInclude`, `CannotOpenImg`. Do you see any reason to not group `ErrorLoadingFile` into `CannotOpenFile` as well?

Reply:
ok @sophstad so we actually changed the name of this error from `ErrorLoadingFile` to `UnmarshallingError` since the issue has to do with the type in the data not matching the python class we are loading it into